e2e/app/component/share-modal.ts

```diff
@@ -0,0 +1,107 @@
+import Container from 'app/container';
+import {ElementHandle, Page} from 'puppeteer';
+import {WorkspaceAccessLevel} from 'app/page-identifiers';
+import {ElementType} from 'app/xpath-options';
+import Button from 'app/element/button';
+import Textbox from 'app/element/textbox';
+import ClrIcon from 'app/element/clr-icon-link';
+
+
+export enum ButtonLabel {
+  Cancel = 'Cancel',
+  Save = 'Save'
+}
+
+const Selector = {
+  defaultDialog: '//*[@role="dialog"]',
+}
+
+export default class ShareModal extends Container {
+
+  constructor(page: Page, xpath: string = Selector.defaultDialog) {
+    super(page, xpath);
+  }
+
+  async shareWithUser(username: string, level: WorkspaceAccessLevel): Promise<void> {
+    const searchBox = await this.waitForSearchBox();
+    await searchBox.type(username);
+
+    const addCollab = await this.waitForAddCollaboratorIcon();
+    await addCollab.click();
+
+    const roleInput = await this.waitForRoleSelectorForUser(username);
+    await roleInput.click();
+
+    const ownerOpt = await this.waitForRoleOption(level);
+    await ownerOpt.click();
+
+    await this.clickButton(ButtonLabel.Save);
+    await this.waitUntilDialogIsClosed();
+  }
+
+  async removeUser(username: string): Promise<void> {
+    const rmCollab = await this.page.waitForXPath(
+      `${this.collabRowXPath(username)}//clr-icon[@shape="minus-circle"]`);
+    await rmCollab.click();
+
+    await this.clickButton(ButtonLabel.Save);
+    await this.waitUntilDialogIsClosed();
+    return;
+  }
+
+  /**
+   * Creates an xpath to a share "row" for a given user within the modal. This
+   * can be combined with a child selector to pull out a control for a user.
+   */
+  private collabRowXPath(username: string): string {
+    // We dip into child contents to find the collab user row element parent.
+    // .//div filters by a relative path to the parent row.
+    return `//*[@data-test-id="collab-user-row" and .//div[` +
+        `@data-test-id="collab-user-email" and contains(text(),"${username}")]]`;
+  }
+
+  async clickButton(buttonLabel: ButtonLabel): Promise<void> {
+    const button = await this.waitForButton(buttonLabel);
+    await button.waitUntilEnabled();
+    return button.click();
+  }
+
+  async waitForButton(buttonLabel: ButtonLabel): Promise<Button> {
+    return Button.findByName(this.page, {containsText: buttonLabel, type: ElementType.Button}, this);
+  }
+
+  async waitForSearchBox(): Promise<Textbox> {
+    return Textbox.findByName(this.page, {name: 'Find Collaborators'}, this);
+  }
+
+  async waitForAddCollaboratorIcon(): Promise<ClrIcon> {
+    return ClrIcon.findByName(this.page, {iconShape: 'plus-circle'}, this);
+  }
+
+  async waitForRoleSelectorForUser(username: string): Promise<Textbox> {
+    const box = new Textbox(
+      this.page, `${this.collabRowXPath(username)}//input[@type="text"]`);
+    await box.waitForXPath({visible: true});
```

Comment:
nit: Line #84 is unnecessary here because the `click()` takes care the finding element too.
```
async click(options?: ClickOptions): Promise<void> {
    return (await this.asElementHandle()).click(options);
  }
```

Reply:
Hmm.. is that obvious? I'm not sure I would have expected that calling `click()` indirectly also waits for the element to become visible.